Reduce memory usage when updating ZipArchives

edwardneal opened the pull request:
There are a handful of different issues surrounding ZipArchive, so this PR touches on a few. It makes ZipArchive more conservative when writing in `WriteFile`. It does this by tracking the type of changes to each archive entry and of the archive itself.

Previously, every ZipArchiveEntry's data would be loaded into memory, the output stream would have its length set to zero, then the archive entries would be written out in sequence. This naturally causes problems when working with very large ZIP files.

I've changed this behaviour by forcing the ZipArchive to track the offset its first deleted entry, `_firstDeletedEntryOffset`. When writing a file:
* Find the offset of the first entry with any changes at all (`startingOffset`), and the first entry with changes to its dynamic-length metadata or entry data (`completeRewriteStartingOffset`).
* If any entries have an offset greater than `startingOffset`, add them to the list (`entriesToWrite`) of entries to persist.
* If any entries have an offset greater than `completeRewriteStartingOffset`, load their contents into memory.
* Position the stream offset at `startingOffset`, then start writing each entry in `entriesToWrite`. When writing:
  * If the entry has changed at all, (or beyond `completeRewriteStartingOffset`) write the header. If not, seek past it.
  * If the entry has changes to its data, write the data (loaded in an earlier step). If not, seek past it.
* If necessary, write (or seek past) each entry's central directory header. Track whether or not any of these were written.
* If any central directory headers were written (or if there are no entries in the archive), write the end-of-central-directory block. If not, seek past it.
* Shrink the stream to its current position.

This relies upon the list of the ZipArchive's existing ZipArchiveEntry records being sorted in offset order, which is now guaranteed on load.

Issue links:
* Resolves #49149: appending a new entry to the ZipArchive is now "free", as long as nothing else has changed in there.
* Contributes to #1543: the original triage was that should limit ourselves to only loading entries which have changes. This PR doesn't do this - it loads the changed entry and everything after it, since the changed entry could easily write data which would extend it into the next entry. I've made an improvement, but I think a complete fix will be more complex (and might look quite like a compacting heap.)
* Resolves #34506: if there are no changes to the ZipArchive, nothing is written.

There's also an issue (1544) which relates to ZipArchiveEntry storing uncompressed data in memory unnecessarily. This PR doesn't fix that issue, but it means that somebody won't fix it and discover that ZipArchive.Dispose is loading the contents into memory anyway.

I've added a number of test cases which cover the corner cases I've thought of, verifying the correct number of writes and the contents of the files as they go. It's quite a core part of ZipArchive though and I'm conscious that a number of libraries depend upon it, so I'm open to adding more.

Edit: From benchmarking the "append case", my results were pretty much as I expected: performance improves, but becomes more variable. Updating/deleting entries at the end of an archive (or after the largest entries in the archive) becomes faster and uses less memory, while updating/deleting entries at the start of an archive takes about as long as it did before.

```

BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3593/23H2/2023Update/SunValley3)
Intel Core i7-8565U CPU 1.80GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.100-preview.3.24204.13
  [Host]     : .NET 9.0.0 (9.0.24.17209), X64 RyuJIT AVX2
  Job-QKCECQ : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

Toolchain=CoreRun 

```
| Benchmark    | Mean    | Error    | StdDev   | Gen0      | Gen1      | Gen2      | Allocated |
|---------- |--------:|---------:|---------:|----------:|----------:|----------:|----------:|
| Pre-PR | 5.128 s | 0.1346 s | 0.3925 s | 1000.0000 | 1000.0000 | 1000.0000 |      2 GB |
| Post-PR | 8.169 ms | 0.2386 ms | 0.7036 ms | | | |      7.24 KB |

edwardneal committed:
Changed behaviour of WriteFile

WriteFile now takes a more granular approach to writing the ZipArchive to its stream.
Adding a new file to the end of the archive will no longer require every file in the archive to be loaded into memory.
Changing fixed-length metadata inside a file will no longer require the entire archive to be re-written.

edwardneal committed:
Added tests for changed WriteFile behaviour

ericstj commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -17,7 +17,7 @@ namespace System.IO.Compression
     {
         private ZipArchive _archive;
-        private readonly bool _originallyInArchive;
+        internal readonly bool _originallyInArchive;
```
I think both these fields could be kept private by making `LocalHeaderOffsetComparer` a nested internal class under `ZipArchiveEntry`.  Nested classes have access to private members.

edwardneal replied:
`_originallyInArchive` and `_offsetOfLocalHeader` are used inside the `ZipArchive` class too, so I've exposed them as internal properties. I've also cleaned up the existing `_versionToExtract` field, which was marked internal and never used outside `ZipArchiveEntry`.

ericstj commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -389,6 +397,10 @@ private set
 
+        // This property's value only relates to the top-level fields of the archive (such as the archive comment.)
+        // New entries in the archive won't change its state.
+        internal DirtyState Dirty { get; private set; }
```
Please don't use the term `Dirty`.  I know it's used sometimes to indicate that something needs to be written but that's unnecessary jargon.  Can you instead choose a name like `ChangedState`.

ericstj reviewed (commented):
I like the premise of making ZipArchive writes more incremental - would like someone from @dotnet/area-system-io-compression to have a look and see if they can give more guidance for direction in what they'd like to see in this PR.  Added a couple comments around what stuck out.

edwardneal committed:
Changes following code review

Renamed Dirty and DirtyState to Changed and ChangeState.
Explicitly assigned Unchanged as a zero-value ChangeState.

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -680,12 +767,36 @@ private void WriteArchiveEpilogue(long startOfCentralDirectory, long sizeOfCentr
+
+        [Flags]
+        internal enum DirtyState
+        {
```
Thanks - done this. There's one place in `ZipArchive.WriteFile` where I continue to use a constant value of 0, and this is because I'm using it to test a bitmask.

edwardneal committed:
Changed field protection levels

Reset _originallyInArchive and _offsetOfLocalHeader to private members, exposed as internal properties.
Also changed _versionToExtract to be private - this isn't ever used in System.IO.Compression outside of ZipArchiveEntry.

ericstj commented:
@carlossanlop can you please review?

udlose commented:
1. is there any update on this?
2. possibly coming in .net 9?

ericstj commented:
It's too late for .NET 9 but we'll have a look at getting into `main` for .NET 10.0.  There's potential fallout from this change and we need to make sure it gets a closer look and plenty of time to bake before going out stable.

edwardneal commented:
Thanks for the update @ericstj.

To expand on the note in the original description:
> I've made an improvement, but I think a complete fix will be more complex (and might look quite like a compacting heap.)

This PR makes an improvement improvement to the average case and resolves the case of appending new entries to the ZIP archive, but it's limited: if the user changes the contents of the first entry in the archive, WriteFile will write the entire archive again - even if the entry content changes have resulted in the entry becoming smaller. The same thing will happen if the user renames the file, even if the encoded name is shorter.

A mechanism similar to a compacting heap would fix this, but I felt that needed more design work. The primary tradeoff which I was thinking about was between unused space in the archives, and the time/IO required to start moving entries around (if that's even possible - ZipArchive deals with unseekable streams.) That tradeoff naturally changes as we move towards the end of an archive, since the IO to append a large entry to the end of an archive is less expensive than the IO which reshuffles enough entries to make free space at the start.

If there's already a data structure which handles this case in .NET and which accounts for the cost of IO, then this PR can use that directly and reduce the risk.

steveharter commented:
From the description's first step:

> Find the offset of the first entry with any changes at all (startingOffset), and the first entry with changes to its dynamic-length metadata or entry data (dynamicDirtyStartingOffset).

I assume then that all of the entries from that first changed entry to the last entry will be re-written and thus have no fragmentation and thus be the same physically on disk as before the PR? If so, then we shouldn't need to add a new enum value to [ZipArchiveMode](https://learn.microsoft.com/dotnet/api/system.io.compression.ziparchivemode) but would just need to update the documentation for `ZipArchiveMode.Append` where it says "The content of the entire archive is held in memory."

steveharter commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -415,7 +427,7 @@ internal void AcquireArchiveStream(ZipArchiveEntry entry)
                 if (!_archiveStreamOwner.EverOpenedForWrite)
                 {
-                    _archiveStreamOwner.WriteAndFinishLocalEntry();
+                    _archiveStreamOwner.WriteAndFinishLocalEntry(true);
```
nit:
```suggestion
                    _archiveStreamOwner.WriteAndFinishLocalEntry(forceWrite: true);
```

steveharter commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -31,6 +31,8 @@ public class ZipArchive : IDisposable
         private byte[] _archiveComment;
         private Encoding? _entryNameAndCommentEncoding;
 
```
nit: unnecessary extra line here

steveharter commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -295,6 +302,12 @@ public long Length
         /// </summary>
         public string Name => ParseFileName(FullName, _versionMadeByPlatform);
 
+        internal ZipArchive.ChangeState Changed { get; private set; }
```
nit: "Changed" -> "Changes"?

"Changed" sounds like a bool, while "Changes" sounds like the flags it is.

steveharter commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:
```diff
@@ -524,42 +537,55 @@ internal void WriteCentralDirectoryFileHeader()
-                ZipGenericExtraField.WriteAllBlocks(_cdUnknownExtraFields, _archive.ArchiveStream);
+            if (_originallyInArchive && Changed == ZipArchive.ChangeState.Unchanged && !forceWrite)
+            {
+                long centralDirectoryHeaderLength = 46
```
nit: add a const for `46` perhaps with the +'s for readability.

steveharter commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -633,41 +653,108 @@ private void WriteFile()
             Debug.Assert(_readEntries);
 
+            // Entries starting after this offset have had a dynamically-sized change. Everything on or after this point must be rewritten.
+            long dynamicDirtyStartingOffset = 0;
```
"dynamicDirtyStartingOffset" -> "rewriteStartingOffset"?

steveharter commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -633,41 +653,108 @@ private void WriteFile()
+                    }
+                    else
+                    {
+                        entriesToWrite.Add(entry);
```
nit: for understandability of this long method, swap\move `entriesToWrite.Add(entry)` to the beginning of the block above (e.g. `if (!entry.OriginallyInArchive)`)

edwardneal committed:
Merge remote-tracking branch 'upstream/main' into issue-1543-triage

edwardneal committed:
Following code review feedback

* Used named parameter when passing a constant to forceWrite.
* Replaced two magic numbers with constants.
* Small cleanup of ZipArchive.WriteFile for clarity.
* Renamed ZipArchiveEntry.Changed to Changes.

edwardneal commented:
Thanks @steveharter - all of the changes in your review look good, I've rolled them forward. I've also brought the PR up to date, since it pre-dates quite a bit of the zlib changes and the .NET 9 branch.

> I assume then that all of the entries from that first changed entry to the last entry will be re-written and thus have no fragmentation and thus be the same physically on disk as before the PR?

That's correct, for changes to an entry's contents or to its dynamic-length metadata (name/content/etc.) An existing entry's ExternalAttributes or LastWriteTime could be changed, and since these are fixed-length fields ZipArchiveEntry will simply rewrite the header as-is; if the only change to an entry is a fixed-length field, subsequent entries won't be rewritten.

The behaviour of `ZipArchiveMode.Update` is covered by another issue (#1544) with an API proposal to help mitigate this (#101243). However, this PR's loosely related to it. Opening a ZipArchive in Update mode currently triggers two pieces of behaviour:
1. When opening a ZipArchiveEntry's stream, the entire entry is loaded into memory.
2. When disposing the ZipArchive, every entry is loaded into memory and written back out to the stream.

This PR only deals with the second behaviour. The post-PR documentation should be updated - we no longer guarantee that the entire archive will be held in memory. Instead, we guarantee that every entry which we [open](https://learn.microsoft.com/en-gb/dotnet/api/system.io.compression.ziparchiveentry.open?view=net-8.0) (and every entry which follows this in the file) will be loaded into memory. The wording of this is a little tricky though. Perhaps something similar to this?

> When you set the mode to Update, the underlying file or stream must support reading, writing, and seeking. The content of every modified entry (and every entry which follows) in the archive is held in memory, and no data is written to the underlying file or stream until the archive is disposed.

steveharter reviewed (approved):
PTAL area owners @dotnet/area-system-io-compression.

This LGTM. I don't see any potential breaking changes. However, a doc issue should be created as discussed to update the wording around the whole zip will be held in memory.

Thanks @edwardneal

carlossanlop commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -633,41 +652,108 @@ private void WriteFile()
+                        {
+                            // If the pending data to write is fixed-length metadata in the header, there's no need to load the full file for
+                            // inflation and deflation.
+                            if ((entry.Changes & (ChangeState.DynamicLengthMetadata | ChangeState.StoredData)) != 0)
```
The comment above this says `there's no need to load the full file for inflation and deflation`.

Can you tell me where in the old code were we expanding the archive and then compressing it again? I don't see it, and it wasn't happening in [LoadLocalHeaderExtraFieldAndCompressedBytesIfNeeded](https://github.com/dotnet/runtime/blob/a5af0ab77caa5ed7a6844fc5f4f459e5edfe23d3/src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs#L574).

carlossanlop replied:
Ah, right. My brain had a glitch. 😄

carlossanlop reviewed (commented):
Thank you so much for this change, @edwardneal . It's going to improve this code a lot. I'll be curious to see if our perf runs capture any significant perf gains out of this new code that should skip entries that don't need to be rewritten. I'd like to assume our benchmarks touch this code at some point. We'll see. 🤞🏼

I left some comments for you to consider, as well as some questions. Can you please take a look?

edwardneal committed:
Merge branch 'main' into issue-1543-triage

edwardneal committed:
Further code review: initial response

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -502,6 +518,9 @@ private void ReadCentralDirectory()
                 }
 
+                // Sort _entries by each archive entry's position
+                _entries.Sort(ZipArchiveEntry.LocalHeaderOffsetComparer.Instance);
```
Thanks for your review.

This comes partially from the [spec](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT). Section 4.3.2 says:

>4.3.2 Each file placed into a ZIP file MUST be preceded by  a "local file header" record for that file.  Each "local file header" MUST be accompanied by a corresponding "central directory header" record within the central directory section of the ZIP file.

Section 4.4.1.3 says:
>4.4.1.3  The entries in the central directory MAY NOT necessarily be in the same order that files appear in the .ZIP file.

Without the sort, `_entries` will be in the order of the central directory headers within the ZIP file. This works well, as long as the order of the central directory headers matches the order of the local file headers. The sort is designed to handle cases where the ZIP archive could contain entries in the sequence A, B, C, and central directory headers in the sequence A, C, B.

carlossanlop replied:
Thanks for sharing the part of the spec indicating that order cannot be guaranteed.

Still, sorting them would be outside of what the spec indicates. We need to make sure we won't be causing a major perf decrease in cases where we modify zip archives that have been heavily updated.

Some things I would ask to ease our minds:

- Have you seen other zip APIs that sort their entries before iterating them?
- Have you tested this opening a file that has most or all their entries in the wrong order?
- Did you get a chance to run the unit tests locally and see if they were taking longer after this change?

If you can create a big zip file with most or all the entries in the wrong order and then share with me, I can help benchmark it and see if there's a big hit.

edwardneal replied:
I've benchmarked the cost of the sort and taken another look at the point in general.

I've put the full benchmark results below, but these essentially amount to 1.7us for 100 files, 133.8us for 5,000 files, 272us for 10,000 files and 1.68ms for 50,000 files. I didn't run the tests locally after the change, but these results would be too brief to actually change the durations.

More broadly though: we sort the ZipArchiveEntrys by their local file header offset because `WriteFile`'s algorithm for figuring out the right place to start writing to the stream relies on the items being in that order. It's possible to do that without the sort, but I think it'd take multiple iterations over the items - it needs to know which entries follow the modified entry in the ZipArchive. I've not seen any libraries which sort like this directly, although SharpZipLib comes quite close in [`ZipFile.BeginUpdate`](https://github.com/icsharpcode/SharpZipLib/blob/master/src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs#L1602); it sorts the list of archive entry updates, not the entire list of entries in the archive.

For performance purposes, I could potentially only perform the sort if the ZipArchive is in Update mode. There are some cases where users would see entries appear in a different order depending on whether they opened it in Read or Update mode, but that might not be a problem - they're usually going to open it by name.

I could technically also shift the sort into the WriteFile method, only performing it if the ZipArchive is being updated. I'm slightly reluctant to do that because I can see an API proposal for ZipArchive.Flush; sorting the list would mean that the order of entries in the archive would change after the user flushes.

I don't think we've got a unit test where the order of central directory entries is different to the order of local file headers, I'll add that over the next few days.

<details>
<summary>LocalHeaderOffsetComparer performance</summary>

```
BenchmarkDotNet v0.14.0, Windows 11 (10.0.22631.4460/23H2/2023Update/SunValley3)
Intel Core i7-8565U CPU 1.80GHz (Whiskey Lake), 1 CPU, 8 logical and 4 physical cores
.NET SDK 9.0.100-rc.1.24452.12
  [Host]   : .NET 9.0.0 (9.0.24.43107), X64 RyuJIT AVX2
  Baseline : .NET 9.0.0 (9.0.24.43107), X64 RyuJIT AVX2

Job=Baseline  EnvironmentVariables=Empty  Toolchain=.NET 9.0  
InvocationCount=65536  UnrollFactor=16
```
| Method    | NumberOfFiles | RandomOrder | Mean         | Error     | StdDev    | Median       | Ratio | RatioSD | Gen0   | Gen1   | Allocated | Alloc Ratio |
|---------- |-------------- |------------ |-------------:|----------:|----------:|-------------:|------:|--------:|-------:|-------:|----------:|------------:|
| Benchmark | 100           | False       |     1.703 μs | 0.0416 μs | 0.1193 μs |     1.717 μs |  1.01 |    0.10 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 100           | True        |     1.696 μs | 0.0399 μs | 0.1125 μs |     1.697 μs |  1.00 |    0.09 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 5000          | False       |   133.825 μs | 2.0884 μs | 1.7439 μs |   133.406 μs |  1.00 |    0.02 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 5000          | True        |   129.090 μs | 2.5411 μs | 4.5821 μs |   126.966 μs |  1.00 |    0.05 | 0.0153 |      - |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 10000         | False       |   271.985 μs | 4.6164 μs | 4.3182 μs |   270.526 μs |  1.00 |    0.02 | 0.0153 | 0.0153 |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 10000         | True        |   275.864 μs | 5.3996 μs | 7.7440 μs |   271.551 μs |  1.00 |    0.04 | 0.0153 | 0.0153 |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 50000         | False       | 1,681.595 μs | 8.1089 μs | 7.1883 μs | 1,682.396 μs |  1.00 |    0.01 | 0.0153 | 0.0153 |      64 B |        1.00 |
|           |               |             |              |           |           |              |       |         |        |        |           |             |
| Benchmark | 50000         | True        | 1,962.890 μs | 7.8197 μs | 6.9320 μs | 1,963.155 μs |  1.00 |    0.00 | 0.0153 | 0.0153 |      64 B |        1.00 |



</details>

carlossanlop replied:
> For performance purposes, I could potentially only perform the sort if the ZipArchive is in Update mode. There are some cases where users would see entries appear in a different order depending on whether they opened it in Read or Update mode, but that might not be a problem - they're usually going to open it by name.

I like this suggestion. If this new behavior would also affect Write mode, it's not really needed there. I would welcome making this behavior exclusive to Update mode. Can you please make that change?

> I don't think we've got a unit test where the order of central directory entries is different to the order of local file headers, I'll add that over the next few days.

Thanks, yes. Please add one. Let me know if I can assist here.

edwardneal replied:
Sounds good, thanks @carlossanlop. I've added the condition, and both cases are tested by the `ArchivesInOffsetOrder_UpdateMode` and `ArchivesInCentralDirectoryOrder_ReadMode` methods.

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -633,41 +652,108 @@ private void WriteFile()
             {
-                entry.WriteCentralDirectoryFileHeader();
+                // The central directory needs to be rewritten if its position has moved, if there's a new entry in the archive, or if the entry might be different.
+                bool centralDirectoryEntryRequiresUpdate = startOfCentralDirectory != _centralDirectoryStart
```
Thanks - I agree it's confusing. I've renamed `startOfCentralDirectory` to `plannedCentralDirectoryPosition`. The instance variable `_centralDirectoryStart` contains the current position of the central directory.

carlossanlop replied:
I'd prefer `plannedStartOfCentralDirectory`. Or maybe `newStartOfCentralDirectory` would be better: shorter and still indicates it will be the new value. Would you agree?

carlossanlop replied:
I'm ok either way. I see you already pushed the commit with `plannedCentralDirectoryPosition `, which is good too.

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -633,41 +652,108 @@ private void WriteFile()
-                entry.WriteCentralDirectoryFileHeader();
+                // The central directory needs to be rewritten if its position has moved, if there's a new entry in the archive, or if the entry might be different.
+                bool centralDirectoryEntryRequiresUpdate = startOfCentralDirectory != _centralDirectoryStart
+                    | (!entry.OriginallyInArchive || (entry.OriginallyInArchive && entry.OffsetOfLocalHeader >= completeRewriteStartingOffset));
```
I've tested with SharpLab and agree. Thanks!

edwardneal commented on src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -633,41 +652,108 @@ private void WriteFile()
+
+                        // We want to re-write entries which are after the starting offset of the first entry which has pending data to write.
+                        // NB: the existing ZipArchiveEntries are sorted in _entries by their position ascending.
+                        if (entry.OffsetOfLocalHeader >= startingOffset)
+                        {
```
We could do, yes - once `entry.OffsetOfLocalHeader` is greater than `startingOffset`, we know that every entry afterwards will trigger the same criteria. Even though the value of `startingOffset` can change, it'll only ever reduce.

I think this comparison would thus become:
```csharp
if (beginRewritingEntries || entry.OffsetOfLocalHeader >= startingOffset)
{
    beginRewritingEntries = true;
   // ...
```

Is this what you have in mind?

carlossanlop replied:
Hmm on a second thought, I think the code is good as it is now. No need to change this. The condition wouldn't help much.

edwardneal committed:
Further code review comments

The list of entries in a ZipArchive is now only sorted when opened in Update mode.
Added/modified a test to verify that these entries appear in the correct order: offset order when opened in Update mode, central directory entry order when opened in Read mode.

carlossanlop committed:
Merge branch 'main' into issue-1543-triage

carlossanlop commented:
Thanks for your latest changes. I don't have any more new feedback _but_ I'm still consulting with some folks on ideas about that sort. Meanwhile I'll run some extra CI runs to see if it all looks good on mobile platforms. I'll come back to you when I get some more answers.

I mentioned this in your other PR, but mentioning it here too: we can try to aim for Preview1 to merge this change. Code Complete day is Monday January 27th. One thing to keep in mind is that I expect a bunch of merge conflicts with your other PR, so let's get ready for that too.

carlossanlop commented:
/azp run runtime-extra-platforms

edwardneal commented:
Thanks carlossanlop. I've checked the logs for the extra runs - there are failures in the same area on the `linux-x64 Release Libraries_Release_CoreCLR` leg, but it looks like this is related to a missing glibc version:
* [System.IO.Compression.Tests](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-102704-merge-81ef42fdc69343e3b3/System.IO.Compression.Tests/1/console.e357ba31.log?helixlogtype=result)
* [System.IO.Compression.ZipFile.Tests](https://helixr1107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-102704-merge-81ef42fdc69343e3b3/System.IO.Compression.ZipFile.Tests/1/console.d694bc06.log?helixlogtype=result)

> Failed to load /root/helix/work/correlation/shared/Microsoft.NETCore.App/10.0.0/libcoreclr.so, error: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /root/helix/work/correlation/shared/Microsoft.NETCore.App/10.0.0/libcoreclr.so)
Failed to bind to CoreCLR at '/root/helix/work/correlation/shared/Microsoft.NETCore.App/10.0.0/'
Failed to create CoreCLR, HRESULT: 0x80008088

carlossanlop commented:
@edwardneal I talked with our security expert @GrabYourPitchforks about any concerns regarding the sort. Here's the summary of the conversation:

- The sorting does not seem concernig, because a `List<ZipArchiveEntry>` will sort the object references under the covers, not the full objects in memory.
- The sort has a complexity of `O(n log n)` as documented in the Remarks section of [`Sort(Comparion<T>)`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.sort?view=net-9.0#system-collections-generic-list-1-sort(system-comparison((-0)))).
- The sort is unstable, meaning that if two elements are equal, their order might not get preserved (also documented in Remarks). But since we're sorting unique numbers (the offset of each entry in the header), we don't have to worry about this limitation.

-----

We did discuss a behavior that needs verification: if `EnsureCentralDirectoryRead` is only called when the archive is opened, then `_readEntries` will only contain the entries that were originally in the archive. What would happen if we open the archive in Update mode, modify the variable-size fields of a few existing entries, but also add some new entries?

I see that in the proposed code, the sort happens in `EnsureCentralDirectoryRead` (existing entries), which means that any added entries will get appended at the end of `_readEntries` and they _should_ get written at the end of the archive.

Unfortunately, there's no test verifying this behavior: opening in Update mode, modifying entries, and adding entries. Can you please add a test for this?

Aside from that final request, and running runtime-extra-platforms again, I think we would be good to merge.

carlossanlop commented:
Can you please merge conflicts since I just merged your other PR? 😄

edwardneal commented:
Thanks both for the review - that sounds good to me! I'll address the merge conflicts today and add the extra test.

edwardneal committed:
Merged main

edwardneal committed:
Correct the write counts in tests

This accounts for the removal of BinaryReader in an earlier PR

edwardneal committed:
Extra test: Update_PerformMinimalWritesWhenEntriesModifiedAndAdded

This test modifies an entry at a specific index, adds N entries after it, then verifies that they've been written in the expected order: existing entries first, new entries afterwards

carlossanlop commented:
/azp run runtime-extra-platforms

carlossanlop reviewed (approved):
LGTM. Assuming no {related} CI issues, we can merge. Thanks @edwardneal!

edwardneal commented:
That's great, thanks for your reviews @carlossanlop. I've checked the test results and they look unrelated to me:
* [StringTests.StartsWithNoMatch_StringComparison](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/Tests/System/StringTests.cs#L7311) on tvOS
* [Shake128Tests](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography/tests/Shake128Tests.cs), [Shake256Tests](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography/tests/Shake256Tests.cs) on Azure Linux
* Regular exit codes of 137 due to an incorrect version of glibc
